internal/controllers/import_controller_v3.go

```diff
@@ -0,0 +1,393 @@
+// +kubebuilder:rbac:groups=management.cattle.io,resources=clusters;clusterregistrationtokens;clusterregistrationtokens/status,verbs=get;list;watch
+
+// Reconcile reconciles a CAPI cluster, creating a Rancher cluster if needed and applying the import manifests.
+func (r *CAPIImportManagementV3Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, reterr error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling CAPI cluster")
+
+	capiCluster := &clusterv1.Cluster{}
+	if err := r.Client.Get(ctx, req.NamespacedName, capiCluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{Requeue: true}, nil
+		}
+
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	if capiCluster.ObjectMeta.DeletionTimestamp.IsZero() && !turtlesannotations.HasClusterImportAnnotation(capiCluster) {
+		if !controllerutil.ContainsFinalizer(capiCluster, managementv3.CapiClusterFinalizer) {
+			log.Info("capi cluster is imported, adding finalizer")
+			controllerutil.AddFinalizer(capiCluster, managementv3.CapiClusterFinalizer)
+
+			if err := r.Client.Update(ctx, capiCluster); err != nil {
+				return ctrl.Result{}, fmt.Errorf("error adding finalizer: %w", err)
+			}
+		}
+	}
+
+	log = log.WithValues("cluster", capiCluster.Name)
+
+	// Wait for controlplane to be ready. This should never be false as the predicates
+	// do the filtering.
+	if !capiCluster.Status.ControlPlaneReady && !conditions.IsTrue(capiCluster, clusterv1.ControlPlaneReadyCondition) {
+		log.Info("clusters control plane is not ready, requeue")
+		return ctrl.Result{RequeueAfter: defaultRequeueDuration}, nil
+	}
+
+	// Collect errors as an aggregate to return together after all patches have been performed.
+	var errs []error
+
+	result, err := r.reconcile(ctx, capiCluster)
+	if err != nil {
+		errs = append(errs, fmt.Errorf("error reconciling cluster: %w", err))
+	}
+
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		capiClusterCopy := capiCluster.DeepCopy()
+
+		patchBase := client.MergeFromWithOptions(capiCluster, client.MergeFromWithOptimisticLock{})
+
+		if err := r.Client.Patch(ctx, capiClusterCopy, patchBase); err != nil {
+			errs = append(errs, fmt.Errorf("failed to patch cluster: %w", err))
+		}
+		return nil
+	}); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if len(errs) > 0 {
+		return ctrl.Result{}, errorutils.NewAggregate(errs)
+	}
+
+	return result, nil
+}
+
+func (r *CAPIImportManagementV3Reconciler) reconcile(ctx context.Context, capiCluster *clusterv1.Cluster) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	// placeholder name/namespace must be set to be able to test this but it won't be used.
+	rancherCluster := &managementv3.Cluster{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "c-",
+			Namespace: capiCluster.Namespace,
+			Labels: map[string]string{
+				capiClusterOwner:          capiCluster.Name,
+				capiClusterOwnerNamespace: capiCluster.Namespace,
+			},
+		},
+	}
```

Comment:
```suggestion
	rancherCluster := &managementv3.Cluster{}
```
Just noticed that this is an invalid object, it has `c-` name on it. If this is needed for tests to work, then what tests are testing? We should not put mocks in production code just for testing. Removing this will also help with readability.

Reply:
what's exactly the issue with tests?

Reply:
Just re-tested this and it is not needed anymore, I'll be fixing it. Honestly I couldn't reproduce the scenario that caused this issue, but I remember tests were complaining about the resource not having name nor namespace and that's why I added a placeholder. Thought it could be coming from when we tried to use resource ownership to link capi cluster to rancher cluster, because we then used a `Get` to fetch the cluster instead of a `List`, but it doesn't seem like that's the case. Thanks for reporting.